### ISSUE TITLE
NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE false positive when add edu.umd.cs.findbugs.annotations.Nullable 

### ISSUE THREAD

**Issue by komuro-hiraku** [spotbugs/spotbugs#3243]
Hello, I have found an issue that appears to be a bug.

## Description

After updating to SpotBugs 4.8.6, the method annotated with edu.umd.cs.findbugs.annotations.Nullable is incorrectly flagged with NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE as a false positive.

## Java version
-  OpenJDK Runtime Environment Corretto-21.0.1.12.1 (build 21.0.1+12-LTS)

## Change Version

- SpotBugs: 4.8.1 to 4.8.6
 
## minimum example

```java
package com.sample.realworld;

import edu.umd.cs.findbugs.annotations.Nullable;

import javax.annotation.Nonnull;

public class Main {
 
    public static void main(String[] args) {
        var hoge = new Hoge();
        System.out.println(hoge.sample);
    }
 
    public static class Hoge {
        @Nonnull public String sample = getNullable();
 
        @Nullable String getNullable() {
            return null;
        }
    }
}
```

Build script

```groovy
plugins {
    id 'java'
    id("com.github.spotbugs") version "6.0.26"
}
 
group = 'com.sample.realworld'
version = '1.0-SNAPSHOT'
 
repositories {
    mavenCentral()
    gradlePluginPortal()
}
 
dependencies {
    testImplementation platform('org.junit:junit-bom:5.10.0')
    testImplementation 'org.junit.jupiter:junit-jupiter'
 
    implementation 'com.diffplug.spotless:spotless-plugin-gradle:6.22.0'
    implementation 'com.github.spotbugs.snom:spotbugs-gradle-plugin:6.0.22'
    spotbugs("com.github.spotbugs:spotbugs:4.8.6")
//    spotbugs("com.github.spotbugs:spotbugs:4.8.1")
 
    compileOnly "com.github.spotbugs:spotbugs-annotations:${spotbugs.toolVersion.get()}"
}
 
test {
    useJUnitPlatform()
}
```

### Run SpotBugs with version 4.8.1:
 
```bash
./gradlew -Duser.language=en -Duser.country=US clean spotbugsMain
 
BUILD SUCCESSFUL in 1s
3 actionable tasks: 3 executed
```

### Run SpotBugs with version 4.8.6:
 
```bash
./gradlew -Duser.language=en -Duser.country=US clean spotbugsMain
Starting a Gradle Daemon, 1 incompatible Daemon could not be reused, use --status for details
 
> Task :spotbugsMain FAILED
M D NP: Possible null pointer dereference in com.sample.realworld.Main.main(String[]) due to return value of called method  Method invoked at Main.java:[line 13]
 
FAILURE: Build failed with an exception. NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE * What went wrong:
Execution failed for task ':spotbugsMain'.
> A failure occurred while executing com.github.spotbugs.snom.internal.SpotBugsRunnerForHybrid$SpotBugsExecutor
   > Verification failed: SpotBugs ended with exit code 1.
```

## expected behavior

NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE is not detected

Thank you.

**Comment by komuro-hiraku** [comment]
@gtoison Thank you reply for my issue.
Up until version 4.8.3, edu.umd.cs.findbugs.annotations.Nullable maintained the behavior as described in this document 
https://spotbugs.readthedocs.io/en/stable/annotations.html#edu-umd-cs-findbugs-annotations-nullable
 
If a null check is necessary, I think edu.umd.cs.findbugs.annotations.CheckForNull should be used. 
Therefore, I expect the original behavior.

**Comment by iloveeclipse** [comment]
I agree with @komuro-hiraku. After update from 4.8.2 to 4.9.0 we see a huge number of NP warnings due this change in semantics of `Nullable`. Ideally this should be fixed as (at least in our code) everything annotated with `Nullable` shows warnings...

**Comment by gtoison** [comment]
I'm most probably misunderstanding the meaning of `@Nullable` but here's the way I'm reading this code:
```
    public static class Hoge {
        @NonNull
        public String sample = getNullable();

        @Nullable String getNullable() {
            return "someValue";
        }
    }
```

Here `sample` MUST be non null, however we're assigning it the return of a method that is potentially null (_The annotated element could be null under some circumstances_)
So we're potentially assigning a null value into a non null field and I would expect to see a warning.

**Comment by iloveeclipse** [comment]
Our use case is a bit different. We have lot of code that operates on maps that are guaranteed to be populated under some conditions. The oversimplified case is something like below:

```
import java.util.HashMap;
import java.util.Map;

import edu.umd.cs.findbugs.annotations.Nullable;

public class Issue_3243 {

	Map<Integer, Double> data = new HashMap<>();
	// some code that fills the map

	@Nullable
	Double getSomething(int id) {
		return data.get(id);
	}
	 
	double doSomething() {
		double doubleValue = getSomething(42).doubleValue(); // no warning before but now!
		return doubleValue;
	}
}
```
Now we get lot of `NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE` warnings.

**Comment by gtoison** [comment]
I think I see what you mean, but then what's the difference between:
```
	@Nullable
	Double getSomething(int id) {
		return data.get(id);
	}
```
And:
```
	Double getSomething(int id) {
		return data.get(id);
	}
```

Isn't it the default already that when there's no `@Nullable` annotation the return is potentially null?

I was under the impression that when there's annotation, SpotBugs attemps to check if the return is nullable, but there's an annotation it only considers the semantic of the annotation

**Comment by iloveeclipse** [comment]
`edu.umd.cs.findbugs.annotations.Nullable` is special. Its semantic always was confusing for users coming from null annotations developed years later which used `Nullable` as synonym for `CheckForNull`.
